[indvars] Missing variables at Og:

CarlosAlbertoEnciso opened the pull request:
https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:
 ```
4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop(Index);
12 }
13 ...
```
Missing local variable `Index` after loop `Induction Variable Elimination`.
When adding a breakpoint at line `11`, LLDB does not have information
on the variable. But it has info on `Var` and `End`.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop(Index);
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

nikic commented on llvm/lib/Transforms/Scalar/IndVarSimplify.cpp:
```diff
@@ -1931,6 +1932,30 @@ bool IndVarSimplify::run(Loop *L) {
+  // for the induction variable with its final value.
+  if (PHINode *IndVar = L->getInductionVariable(*SE)) {
+    const SCEV *IndVarSCEV = SE->getSCEVAtScope(IndVar, L->getParentLoop());
+    if (IndVarSCEV->getSCEVType() == SCEVTypes::scConstant) {
```
`isa<SCEVConstant>`.

nikic commented on llvm/lib/Transforms/Scalar/IndVarSimplify.cpp:
```diff
@@ -1931,6 +1932,30 @@ bool IndVarSimplify::run(Loop *L) {
   }
 
+  // The loop exit values have been updated; insert the debug location
+  // for the induction variable with its final value.
```
Why is this code not inside rewriteLoopExitValues, which does the actual transform?

CarlosAlbertoEnciso replied:
Moved to `rewriteLoopExitValues`.

nikic commented on llvm/test/Transforms/LoopSimplify/pr51735-1.ll:
```diff
@@ -0,0 +1,115 @@
+; CHECK:   %call = tail call noundef i32 @_Z3nopi(i32 noundef 777), {{.*}}
+; CHECK-DAG: ![[DBG]] = !DILocalVariable(name: "Index"{{.*}})
+
+define dso_local noundef i32 @_Z3nopi(i32 noundef %Param) local_unnamed_addr #0 !dbg !10 {
```
Unrelated function?

CarlosAlbertoEnciso replied:
Removed. (For both test cases).

nikic commented on llvm/test/Transforms/LoopSimplify/pr51735-1.ll:
```diff
@@ -0,0 +1,115 @@
+  ret void, !dbg !35
+}
+
+define dso_local noundef i32 @main() local_unnamed_addr #3 !dbg !36 {
```
Unrelated function?

CarlosAlbertoEnciso replied:
Removed. (For both test cases).

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop(Index);
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

- Moved the new logic to 'rewriteLoopExitValues'.
- Removed from the test cases, the unrelated functions: 'nop' and 'main'.
- Use 'isa<SCEVConstant>'.

CarlosAlbertoEnciso commented:
@nikic Thanks for your feedback.
I uploaded a new patch addressing:
- Moved the new logic to `rewriteLoopExitValues`.
- Removed from the test cases, the unrelated functions: `nop` and `main`.
- Use `isa<SCEVConstant>`.

nikic commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1519,6 +1520,31 @@ int llvm::rewriteLoopExitValues(Loop *L, LoopInfo *LI, TargetLibraryInfo *TLI,
+
+  // The loop exit values have been updated; insert the debug location
+  // for the induction variable with its final value.
+  if (PHINode *IndVar = L->getInductionVariable(*SE)) {
```
This is still the wrong place to do this. You are updating just a single "induction variable" here, while the code can rewrite multiple, and can also use different exit values for different blocks. This adjustment should be above in the `// Transformation.` loop, which has all the induction phis and their replacement values.

felipepiovezan commented on llvm/test/Transforms/LoopSimplify/pr51735.ll:
```diff
@@ -0,0 +1,85 @@
+; CHECK:   call void @llvm.dbg.value(metadata i32 777, metadata ![[DBG:[0-9]+]], {{.*}}
+; CHECK-DAG: ![[DBG]] = !DILocalVariable(name: "Index"{{.*}})
+
+define dso_local void @_Z3barv() local_unnamed_addr #1 !dbg !15 {
```
Could you elaborate on how these tests differ from each other? I couldn't tell by comparing the two

jmorse replied:
(IIRC, the difference is that there are multiple exit blocks, to stimulate more of the paths through the code-changes).

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop();
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

Address reviewers comments.
- Early exit to simplify the logic.
- Avoid inserting the same instruction in multiple blocks.
- Skip debug-users with variadic variable locations.
- Change some comments to improve readability.
- Add code to clone and move the debug value.
- Modify second test case to include multiple exit blocks.

CarlosAlbertoEnciso commented:
I am sorry for my delay in addressing the comments from the reviewers, but I was on PTO.
The reworked patch addresses the issues and includes 2 test cases:
- Multiple exit blocks
- Single exit block.

nikic commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1341,6 +1342,36 @@ static bool checkIsIndPhi(PHINode *Phi, Loop *L, ScalarEvolution *SE,
+    // get updated, which is fine as that is the existing behaviour.
+    if (DebugUser->hasArgList())
+      continue;
+    for (BasicBlock *Exit : ExitBlocks) {
```
The ExitValue is specific to one exit block. Why are you updating all of them here?

CarlosAlbertoEnciso replied:
Reworked the code to take into account that point.

nikic commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1482,6 +1513,21 @@ int llvm::rewriteLoopExitValues(Loop *L, LoopInfo *LI, TargetLibraryInfo *TLI,
           &*Inst->getParent()->getFirstInsertionPt() : Inst;
         RewritePhiSet.emplace_back(PN, i, ExitValue, InsertPt, HighCost);
+
+        // Add debug values if the PN is a induction variable.
```
Why is this code needed in addition to the addDebugValuesToLoopVariable() below?

CarlosAlbertoEnciso replied:
That code creates a debug value in the exit block if any of the incoming values for the PHI Node being replaced is the induction variable. Reworked in new patch.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop();
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

Address reviewers comments.
- Early exit to simplify the logic.
- Avoid inserting the same instruction in multiple blocks.
- Skip debug-users with variadic variable locations.
- Change some comments to improve readability.
- Add code to clone and move the debug value.
- Modify second test case to include multiple exit blocks.

Addressed the upstream feedback in relation to:
- Each exit block has its own exit value.
- Separate the debug values for incoming and exit values.

CarlosAlbertoEnciso commented:
Uploaded a new patch that addresses @nikic feedback.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Fix clang-format found issues in updated patch.

CarlosAlbertoEnciso commented:
Uploaded patch fixing the C/C++ code formatter, clang-format found issues.

jmorse commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1540,11 +1590,22 @@ int llvm::rewriteLoopExitValues(Loop *L, LoopInfo *LI, TargetLibraryInfo *TLI,
   }
 
+  // If there are no PHIs to be rewritten then there are no loop live-out
+  // values, try to rewrite variables corresponding to the induction variable
```
I reckon you want to say "rewrite debug variables" or similar, just to signal to the reader that these aren't for SSA variables which might have a functional effect on the program.

CarlosAlbertoEnciso replied:
Changed.

jmorse commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1540,11 +1590,22 @@ int llvm::rewriteLoopExitValues(Loop *L, LoopInfo *LI, TargetLibraryInfo *TLI,
+  // will completely forget that this loop happened.
+  if (RewritePhiSet.empty()) {
+    // The loop exit value has been updated; insert the debug location
+    // for the given the induction variable with its final value.
```
"for the given induction..."

CarlosAlbertoEnciso replied:
Changed.

jmorse reviewed (commented):
For the change adding the "ExitBlock" field to the RewritePhi struct -- as far as understand it, the PHI being rewritten is always going to be in that exit block. Perhaps we could not store it in the struct and instead recover the exit block by calling `getParent()` on the PHI when necessary, instead? (This avoids two storing two pieces of information and having to think about whether they're in sync or not).

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Addressed the upstream feedback in relation to:
- Remove the introduced 'ExitBlock' field.
- Update some comments.

CarlosAlbertoEnciso commented:
> For the change adding the "ExitBlock" field to the RewritePhi struct -- as far as understand it, the PHI being rewritten is always going to be in that exit block. Perhaps we could not store it in the struct and instead recover the exit block by calling `getParent()` on the PHI when necessary, instead? (This avoids two storing two pieces of information and having to think about whether they're in sync or not).

Good point. Removed the added field.

jmorse reviewed (approved):
This now LGTM, with a need to elaborate some of the documentation before landing. It's also best to leave this a day or two before landing just in case there's any further review feedback.

nikic commented on llvm/test/Transforms/LoopSimplify/pr51735-1.ll:
```diff
@@ -0,0 +1,120 @@
+; RUN: opt -passes=indvars -S -o - < %s | FileCheck %s
```
These tests should be in the IndVarSimplify directory, not the LoopSimplify directory.

nikic reviewed (commented):
This patch still looks pretty odd to me.

The part where we add the debug value if the exit value is rewritten (part of RewritePhiSet) makes sense to me.

The rest does not, for a couple of reasons:

 * The use of getInductionVariable() is always a good indicator that you are doing something dubious: It will only recognize some "canonical" induction variable, but a loop may have many IVs. So it looks like you will be preserving the exit value of a single IV only?
 * You are adding debug values in a case where *the transform in question does not actually fire*. This indicates to me that you are doing this in the wrong place. I believe that for the case where the IV is effectively dead, the place where it actually gets dropped is in DeleteDeadPHIs, so I would expect that debuginfo preservation would also happen in that area.
 * If I'm reading this correctly, for the same reason as the previous point, you will be adding these debug values even if the IV does *not* actually get dropped (because it has real uses in the loop) -- I don't think that's supposed to happen?

Disclaimer: I know nothing about debuginfo, maybe I'm misunderstanding things.

jmorse commented:
Hmmm -- this started out as a patch that was to be for loop-deletion, it looks like I missed that this was firing at all times during indvars, that's my bad. The original rationale for all of this was that when loops are completely erased, we lose everything to do with debug-info that happened in the loop, which has the potential to mislead the developer. With that in mind, I believe the patch is collecting the correct information, just it's installing it at the wrong time. We should thus (IMO) be storing that information until loop-deletion commits to deleting a loop, at which point we definitely need to add intrinsics to the exit blocks.

Just to zip through Nikita's comments,

> you will be adding these debug values even if the IV does not actually get dropped

Yup, that's definitely broken,

> It will only recognize some "canonical" induction variable, but a loop may have many IVs.

Alas yes; as debug-info is to a large extent "best effort", we don't guarantee that it's always 100% completely correct, and there are some scenarios where we just bail out and say "that's too hard to solve". Seeing how deleting intrinsics is always unsafe (and is what happens on loop deletion today), it's better to have one patch that covers 90% of scenarios now than one that covers 100% later.

> If I'm reading this correctly, for the same reason as the previous point, you will be adding these debug values even if the IV does not actually get dropped

Assuming this is the `RewritePhiSet.empty()` portion -- there might be no users of the induction variable outside the loop, but we still have a debug-info interest in describing the exit value after the loop exits. Although: only if the loop and it's intrinsics get deleted.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

After internal discussion with @jmorse, it was decided
to split the work between the 'indvars' and the
'loop-deletion' passes.

1) passes="loop(indvars)"
- 'indvars' transformation is fired:
  the 'rewriteLoopExitValues' will rewrite the collected
  PNs with the exit values.
- 'indvars' transformation is not fired:
  If the loop can be deleted, we preserve the induction
  variable information to be used by 'loop-deletion' if
  that pass will be executed.

2) passes="loop(indvars,loop-deletion)"
  If the loop is deleted in 'deleteDeadLoop' and there
  is a valid exit block, we use any collected values
  by 'indvars' to updated the exit values.

Added extra tests to cover the following cases:
  ...
  char Var = 1;
  for (; Index < End; ++Index)
    if (Index == 666)
      ++Var;
  ...

and
  ...
  char Var = 1;
  for (; Index < End; ++Index)
    if (Index == 666)
      Var = 555;
  ...

Modified but otherwise unused variable 'Var' in a loop
that gets deleted.

SLTozer commented on llvm/test/Transforms/IndVarSimplify/pr51735-2.ll:
```diff
@@ -0,0 +1,129 @@
+; PRE-CHECK:   br label %for.cond
+
+; PRE-CHECK: for.end:
+; PRE-CHECK-NOT: call void @llvm.dbg.value
```
This can be skipped since there's an implicit-check-not.

CarlosAlbertoEnciso replied:
Removed.

SLTozer commented on llvm/test/Transforms/IndVarSimplify/pr51735-2.ll:
```diff
@@ -0,0 +1,129 @@
+
+; Only the 'indvars' pass is executed.
+; PRE-CHECK: for.cond:
+; PRE-CHECK:   call void @llvm.dbg.value(metadata i32 %Var.0, metadata ![[DBG]], {{.*}}
```
This and a lot of the other check lines may fail in a build without assertions enabled, as the SSA values won't necessarily be printed with the same names (they may become numbered values, e.g. `%5`, `%6`, etc). Usually the way to test these would then be to use capture variables for them, e.g. `i32 %[[SSA_VAR0:.+]],`. This gets tedious if you have to use it for every variable, so I'd recommend removing the unrelated CHECK lines, which would be most of the lines that don't need to be tested for and don't produce a value that we want to check; or instead of removing them, you could cut out actual SSA values, i.e. you could just test for `= icmp eq`.

CarlosAlbertoEnciso replied:
Thanks for the additional information. Changed to use `captured variables`.

SLTozer commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1412,6 +1424,36 @@ static bool checkIsIndPhi(PHINode *Phi, Loop *L, ScalarEvolution *SE,
+      continue;
+    auto *Cloned = cast<DbgVariableIntrinsic>(DebugUser->clone());
+    Cloned->replaceVariableLocationOp(0u, PN);
+    Cloned->insertBefore(Successor->getFirstNonPHI());
```
We should fix this up for the new insert logic (same in the function below):
```suggestion
    Cloned->insertBefore(Successor, Successor->getFirstNonPHIIt());
```

CarlosAlbertoEnciso replied:
Changed to use new insert logic.

SLTozer reviewed (commented):
Largely looks good, some comments inline.

CarlosAlbertoEnciso commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -607,6 +608,17 @@ void llvm::deleteDeadLoop(Loop *L, DominatorTree *DT, ScalarEvolution *SE,
+      SmallVector<WeakVH> &DbgUsers = L->getDebugInductionVariableDebugUsers();
+      for (WeakVH &DebugUser : DbgUsers)
+        if (DebugUser)
+          dyn_cast<DbgVariableIntrinsic>(DebugUser)->replaceVariableLocationOp(
```
Thanks for the additional information on `dyn_cast`. Changed to `cast`.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Address @SLTozer comments.
- Use captured variables for SSA values.
- Use new insert logic calls.
- Remove not required 'CHECK-NOT'.

CarlosAlbertoEnciso commented:
@SLTozer Thanks for your comments. Updated the patch.

SLTozer commented:
One more part - you've captured the SSA values that are needed for the test, but for the SSA values you're not using you should still remove the actual names so that if the identifiers change (which can happen when building with assertions enabled I think) the test doesn't fail. Since you don't need to use the captured names, you can just use `{{.+}}`, or else you can just omit some of the lines in whole or in part.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Address @SLTozer comments.
- Don't use any names for not used SSA values.

CarlosAlbertoEnciso commented:
@SLTozer Thanks for the suggested changes when dealing with the not used SSA values.

CarlosAlbertoEnciso commented:
Any additional comments or suggestions? Thanks.

CarlosAlbertoEnciso commented:
Thanks to all reviewers for their comments.

nikic commented:
This introduces a layering violation. GenericLoopInfo in Support is not allowed to depend on IR. That's the point of the GenericXYZ implementations, they are IR-agnostic.

nikic commented on llvm/include/llvm/Support/GenericLoopInfo.h:
```diff
@@ -75,6 +77,13 @@ template <class BlockT, class LoopT> class LoopBase {
+  // there are no PHIs to be rewritten.
+  // For now, we only preserve single induction variables.
+  Value *IndVarFinalValue = nullptr;
+  SmallVector<WeakVH> IndVarDebugUsers;
```
Why is it necessary for these to be members on the Loop object itself?

CarlosAlbertoEnciso replied:
It also introduced some link errors. Reverting the patch.

CarlosAlbertoEnciso replied:
Thanks for reverting. I will look at that layering violation.

CarlosAlbertoEnciso commented:
Reopen as https://github.com/llvm/llvm-project/pull/88270
